org.eclipse.winery.topologymodeler.ui/src/app/canvas/canvas.component.ts

```diff
@@ -0,0 +1,1403 @@
+            typeQName: this.policyTypeQName,
+        };
+        this.ngRedux.dispatch(this.actions.setPolicy(policies));
+        this.cancel('cancelPolicies');
+    }
+
+    /**
+     * Auto-completes other policy relevant values when a policy template has been selected in the modal
+     */
+    onChangePolicyTemplate(policyTemplate: string): void {
+        this.entityTypes.policyTemplates.map(ct => {
+            if (ct.id === policyTemplate) {
+                this.policyTemplateColor = ct.color;
+                this.policyTemplateId = ct.id;
+                this.policyTemplateName = ct.name;
+                this.policyTemplateNamespace = ct.namespace;
+                this.policyTemplateQName = ct.qName;
+            }
+        });
+    }
+
+    /**
+     * Auto-completes other policy relevant values when a policy type has been selected in the modal
+     */
+    onChangePolicyType(policyType: string): void {
+        this.entityTypes.policyTypes.map(ct => {
+            if (ct.id === policyType) {
+                this.policyTypeColor = ct.color;
+                this.policyTypeId = ct.id;
+                this.policyTypeName = ct.name;
+                this.policyTypeNamespace = ct.namespace;
+                this.policyTypeQName = ct.qName;
+            }
+        });
+    }
+
+    /**
+     * Clears the modal values when clicking on the cancel button
+     */
+    cancel($event): void {
+        let modalType;
+        try {
+            modalType = $event.srcElement.id;
+        } catch (e) {
+            modalType = $event;
+        }
+        switch (modalType) {
+            case 'cancelPolicies':
+                this.policyTemplateName = null;
+                this.policyType = null;
+                this.policyTemplate = null;
+                this.policiesModal.hide();
+                break;
+            case 'cancelRequirements':
+                this.reqId = null;
+                this.reqDefinitionName = null;
+                this.reqType = null;
+                this.requirementsModal.hide();
+                break;
+            case 'cancelCapabilities':
+                this.capId = null;
+                this.capDefinitionName = null;
+                this.capType = null;
+                this.capabilitiesModal.hide();
+                break;
+            case 'cancelDeploymentArtifacts':
+                this.artifactTemplateNS = null;
+                this.artifactTemplateName = null;
+                this.artifactName = null;
+                this.artifactType = null;
+                this.deploymentArtifactModal.hide();
+                break;
+        }
+    }
+
+    /**
+     * New nodes can be dragged directly from the palette,
+     * adds the node to the internal representation
+     * @param event  The html event.
+     */
+    moveNewNode(event): void {
+        this.newNodeData = {
+            id: this.newNode.id,
+            x: event.clientX - 108,
```

Comment:
because the mouse pointer should be in the middle of the new generated node, this is needed for each node as an offset otherwise the mouse pointer would be in the bottom right corner upon creation